src/app/InteractionModelEngine.cpp

```diff
@@ -240,9 +240,78 @@ DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aComman
         "Default DispatchSingleClusterCommand is called, this should be replaced by actual dispatched for cluster commands");
 }
 
+CHIP_ERROR __attribute__((weak))
+ReadSingleClusterData(NodeId aNodeId, ClusterId aClusterId, EndpointId aEndPointId, FieldId aFieldId, TLV::TLVWriter & aWriter)
+{
+    ChipLogDetail(DataManagement,
+                  "Received Cluster Command: Cluster=%" PRIx16 " NodeId=%" PRIx64 " Endpoint=%" PRIx8 " FieldId=%" PRIx8,
+                  aClusterId, aNodeId, aEndPointId, aFieldId);
+    ChipLogError(DataManagement,
+                 "Default ReadSingleClusterData is called, this should be replaced by actual dispatched for cluster");
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR __attribute__((weak))
+WriteSingleClusterData(NodeId aNodeId, ClusterId aClusterId, EndpointId aEndPointId, FieldId aFieldId, TLV::TLVReader & aReader)
+{
+    ChipLogDetail(DataManagement,
+                  "Received Cluster Command: Cluster=%" PRIx16 " NodeId=%" PRIx64 " Endpoint=%" PRIx8 " FieldId=%" PRIx8,
+                  aClusterId, aNodeId, aEndPointId, aFieldId);
+    ChipLogError(DataManagement,
+                 "Default WriteSingleClusterData is called, this should be replaced by actual dispatched for cluster");
+    return CHIP_NO_ERROR;
+}
+
 uint16_t InteractionModelEngine::GetReadClientArrayIndex(const ReadClient * const apReadClient) const
 {
     return static_cast<uint16_t>(apReadClient - mReadClients);
 }
+
+CHIP_ERROR InteractionModelEngine::GetFirstAvailableClusterInfo(ClusterInfo *& apClusterInfo)
+{
+    if (mNumClusterInfos >= IM_SERVER_MAX_NUM_PATH_GROUPS)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    apClusterInfo = &mClusterInfoPool[mNumClusterInfos];
+    mNumClusterInfos++;
+    apClusterInfo->ClearDirty();
+    return CHIP_NO_ERROR;
+}
+
+void InteractionModelEngine::ReleaseClusterInfoList(ClusterInfo * apClusterInfo)
+{
+    ClusterInfo * clusterInfo = apClusterInfo;
```

Comment:
how is the release recorded?

I see this setting all elements dirty and next to null.

Should we not have some sort of free list? or is the "Dirty" flag indicating in use? If so, could we rename it as  such?

Reply:
dirty means that the changed need to be sent out in report. just updated, release would release the cluster info list back to the front of the pool, and update "free list start" accordingly. thanks